### PR TITLE
feat: add createResource endpoint [DEV-4605][DEV-4636]

### PR DESCRIPTION
**Pull request by sownak**
- add PUT /createResource endpoint
- add GET /traits endpoint
- add GET /methods endpoint
- add GET /properties endpoint
- deprecate PUT /create-resource endpoint

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-4605 Update Cheqd registrar /createResource operation](https://app.clickup.com/t/6600954/DEV-4605)

**Comment by ankurdotb** [comment]
Task linked: [DEV-4636 Add helper functions for did-registrar](https://app.clickup.com/t/6600954/DEV-4636)

File: src/controllers/resource.ts [hunk 1]
```diff
@@ -29,13 +29,29 @@ export class ResourceController {
...
+		check('jobId')
+			.custom((value, { req }) => {
+				if (!value && !(req.body.name && req.body.type && req.body.content)) return false;
+				return true;
+			})
+			.withMessage('name, type and content are required'),
+		check('name').optional().isString().withMessage(Messages.Invalid),
+		check('type').optional().isString().withMessage(Messages.Invalid),
+		check('content').exists().isString().withMessage(Messages.Invalid),
+		check('relativeDidUrl').optional().isString().contains('/resources/').withMessage(Messages.InvalidDidUrl),
```
Comment (DaevMithran) [inline src/controllers/resource.ts:43]: where is `relativeDidUrl` used in createResource?

**Comment by DaevMithran** [comment]
Great work!!

File: src/controllers/resource.ts [hunk 2]
```diff
@@ -29,13 +29,29 @@ export class ResourceController {
...
+		check('did').exists().isString().contains('did:cheqd').withMessage(Messages.InvalidDid),
+		check('jobId')
+			.custom((value, { req }) => {
+				if (!value && !(req.body.name && req.body.type && req.body.content)) return false;
+				return true;
+			})
+			.withMessage('name, type and content are required'),
+		check('name').optional().isString().withMessage(Messages.Invalid),
+		check('type').optional().isString().withMessage(Messages.Invalid),
+		check('content').exists().isString().withMessage(Messages.Invalid),
```
Comment (DaevMithran) [inline src/controllers/resource.ts:42]: content can be optional, it's not required when the user is submitting a signatureResponse. It is only required when jobId is not required. Which is already handled above

File: src/controllers/resource.ts [hunk 3]
```diff
@@ -139,4 +157,114 @@ export class ResourceController {
...
+			// check if did is registered on the ledger
+			let resolvedDocument = await CheqdResolver(did);
+			if (!resolvedDocument?.didDocument || resolvedDocument.didDocumentMetadata.deactivated) {
+				return response
+					.status(400)
+					.send(Responses.GetInvalidResponse({ id: did }, secret, Messages.DidNotFound));
+			} else {
+				resolvedDocument = resolvedDocument.didDocument;
+			}
+
```
Comment (DaevMithran) [inline src/controllers/resource.ts:190]: To discuss: should we add an additional check to if the resource name and type already exist. That would prevent someone from updating a resource using the create endpoint.

Comment (sownak) [reply src/controllers/resource.ts:190]: Lets block update-resource via createResource.

File: tests/did/validateDid.spec.ts
Comment (Eengineer1) [inline tests/did/validateDid.spec.ts:1]: I'd suggest splitting the positive and negative tests in two self-explanatory files for better readability and context retention.

File: tests/resource/createResource.spec.ts
Comment (Eengineer1) [inline tests/resource/createResource.spec.ts:1]: Same applies as above.

File: tests/resource/updateResource.spec.ts
Comment (Eengineer1) [inline tests/resource/updateResource.spec.ts:1]: Same applies as above.

File: tests/resource/validateResource.spec.ts
Comment (Eengineer1) [inline tests/resource/validateResource.spec.ts:1]: Same applies as above.

**Review by Eengineer1 — CHANGES_REQUESTED** [review]

File: tests/did/validateDid.spec.ts
Comment (sownak) [reply tests/did/validateDid.spec.ts:1]: All validateTests are negative test cases, they will fail. In this file there is one case to initialise a valid didDoc

**Review by Eengineer1 — DISMISSED** [review]
Looks great!

Approved.

**Review by Eengineer1 — APPROVED** [review]

**Pull request merged**